### PR TITLE
Allow users to choose between AESCBC and AESGCM when configuring envelope encryption.

### PR DESCRIPTION
**Pull request by immutableT**
**What type of PR is this?**

/kind api-change
/kind feature

**What this PR does / why we need it**:
This PR allows users to choose AESGCM protocol (as opposed to currently set by default AESCBC) while configuring KMS Provider.
**Which issue(s) this PR fixes**:

Fixes #
#81127
**Special notes for your reviewer**:
I will follow-up with a PR to update documentation.

**Does this PR introduce a user-facing change?**:

```release-note
Users will be able to supply either AESGCM or AESCBC as the data encryption algorithm.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
if a user specifies AESGCM in this field in the config to an older API server, it will ignore the field because it doesn't understand it, and just encrypt/decrypt with CBC, right? then, on upgrade to 1.18, it will switch to expecting data encrypted with AESGCM and fail to read all existing data, correct?

**Comment by immutableT** [comment]
@liggitt I believe that you are correct.

Some ideas on how to handle such a case:
1. Produce a better error message that will explain what may have happened with instructions on how to fix the issue.
2. Always send both CBC and GCM transformers to the gRPCService, where the top transformer will be the one provided by the config. While transforming from storage try the second transformer when a specific error is encountered. 
3. Examine the raw encrypted message and determine the appropriate transformer to use (this still implies that we send two transformers to gRPCService).

File: test/integration/master/kms_transformation_test.go [hunk 1]
```diff
@@ -111,74 +119,115 @@ resources:
...
+			dekPlainAsWouldBeSeenByETCD := decryptResponse.Plain
+
+			if !bytes.Equal(plainTextDEK, dekPlainAsWouldBeSeenByETCD) {
+				t.Fatalf("expected plainTextDEK %v to be passed to KMS Plugin, but got %s",
+					plainTextDEK, dekPlainAsWouldBeSeenByETCD)
+			}
+
+			plainSecret, err := envelope.plainTextPayload(secretETCDPath, tt.transformerFunc)
+			if err != nil {
+				t.Fatalf("failed to transform from storage via AESCBC, err: %v", err)
```
Comment (awly) [inline test/integration/master/kms_transformation_test.go]: Update the message

**Comment by liggitt** [comment]
> While transforming from storage try the second transformer when a specific error is encountered.

That seems plausible for two encryption methods. Does it become problematic if more are added in the future?

> Examine the raw encrypted message and determine the appropriate transformer to use (this still implies that we send two transformers to gRPCService).

Do we include anything in the encrypted content today that would identify it was encrypted with AESCBC?

Would be interested to hear @smarterclayton's thoughts.

File: test/integration/master/kms_transformation_test.go (hunk 1 above)
Comment (immutableT) [reply test/integration/master/kms_transformation_test.go]: Done.

**Review by immutableT — COMMENTED** [review]
I would suggest to only add this second provider and only for some fixed period of time (ex. three minor versions). At the point when the version that handles AESGCM will been out for a while I think that we should remove that second provider.

No. Today, we don't add any metadata (describing the encryption protocols) to the encrypted payload. I feel that we should do this in the future. 
What I meant by "examining the raw encrypted message" is looking for (hopefully present) patterns of how CBC vs. GCM packages their payloads - quite fragile.

After thinking more about this, I believe that the trade-off comes down to:
fixing the perception that currently exists within security community that the current implementation of the envelope encryption feature is not following the best practices
vs. 
breaking some users (temporarily) who may have misread our documentation and enabled GCM mode on a k8s version that does not yet support it.

I am in favor, for addressing 1 since I believe that there may be a segment of customers that may refrain from using this features just because of this perception. 
Furthermore, (assuming that we improve the error message) fixing the issue caused by 2  is fairly simple (just remove the DataEncryptionProtocol field).

**Comment by liggitt** [comment]
> I would suggest to only add this second provider and only for some fixed period of time (ex. three minor versions). At the point when the version that handles AESGCM will been out for a while I think that we should remove that second provider.

What is that time period based on? What makes it safe to remove after a few releases?

**Comment by enj** [comment]
I have opened #86124 to attempt to fix this without adding config options (I need to fix it to work correctly with HA on downgrade and upgrade).  I think it is unfair to expect users to understand what option to use, and to migration from one option to the other.

**Pull request closed without merging**